API changes (node's border and edge)

EdoFro opened the pull request:
## Changes:
1. expose Border object and its color methods
2. expose isColorSet for Edge object

## Why?
- I wanted to learn how to expose internal methods through Freeplane's API
- I want to make an add-on that needs to access the nodes edges and border colors and if they were set in the node or not.

## Pending
- there are other methods related to borders that may be exposed in future changes (like line width, line type, etc), but I didn't need them this time. Maybe for next time.

Regards,

Edo

EdoFro committed:
API changes

1. expose Border object and its color methods
2. expose isColorSet for Edge object

dpolivaev commented on freeplane_plugin_script/src/main/java/org/freeplane/plugin/script/proxy/BorderProxy.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    public boolean isColorSet() {
+        return NodeBorderModel.getModel(getDelegate())==null?false:(NodeBorderModel.getModel(getDelegate()).getBorderColor() != null);
```
The delegate can not be null. Why do you check it?

EdoFro replied:
I check if there is a Model assigned to that delegate.

Like this:
https://github.com/freeplane/freeplane/blob/6bf842791d968c9e2a35c4432627d37e2a1ac9d2/freeplane/src/main/java/org/freeplane/features/nodestyle/NodeBorderModel.java#L62-L65

where:
https://github.com/freeplane/freeplane/blob/6bf842791d968c9e2a35c4432627d37e2a1ac9d2/freeplane/src/main/java/org/freeplane/features/nodestyle/NodeBorderModel.java#L36-L38

-----

As far as I understood, if the node has no border style assigned, it won't have a NodeBorderModel created:

Example:
Node without NodeBorderModel created:
![image](https://github.com/freeplane/freeplane/assets/35700575/85c6052e-2398-440d-9c5f-15ee78aac654)

Node border format was modified for this node --> FP creates a NodeBorderModel for it:
![image](https://github.com/freeplane/freeplane/assets/35700575/29844dd0-7dd5-4872-8cc1-992fe5c934c7)

dpolivaev replied:
getDelegate() which you compare to null is a NodeModel, isn't it?

EdoFro replied:
yes, it is.

I compare the associated NodeBorderModel to null , NOT the delegate (NodeModel)

the code:
```java
return NodeBorderModel.getModel(getDelegate())==null ? false : (Nod....
```

can also be written as:
```java
NodeModel myNode = getDelegate();
NodeBorderModel nodeBorderModel = NodeBorderModel.getModel(myNode);
return nodeBorderModel==null ? false : (Nod....
```

I don't know if I understand your question correctly. I hope I had answered it.

Best regards,

edo

dpolivaev replied:
Yes, then I misunderstood the code.
Thank you for the clarification.
By the way, instead of using 
````
NodeBorderModel.getModel(getDelegate())==null?false:(NodeBorderModel.getModel(getDelegate()).getBorderColor() != null)
````

I would probably write

````
       NodeBorderModel border = NodeBorderModel.getModel(getDelegate());
       return border!=null && border.getBorderColor() != null;
````

to avoid duplication and to simplify the expression.

dpolivaev commented:
org.freeplane.features.nodestyle.NodeBorderModel supports not only color. Why does the new API support only color?

EdoFro commented:
> ## Pending
> * there are other methods related to borders that may be exposed in future changes (like line width, line type, etc), but I didn't need them this time. Maybe for next time.

> org.freeplane.features.nodestyle.NodeBorderModel supports not only color. Why does the new API support only color?

This is the first time I try to modify an API. I don't know if I did it correctly. It works, but maybe I forgot to include some important steps. I wanted to keep it simple for you to review it and for me to understand it.

But if it is correct, I can add the other methods too. No problem. It is very similar to the Edge object, so I can use that as a guide.

dpolivaev commented:
I do not test your submission. I generally give you only feedback on topics that catch my attention. And I trust that if you submit a pull request you have already tested it or warned me if it is not complete.

EdoFro commented:
> I do not test your submission. I generally give you only feedback on topics that catch my attention. And I trust that if you submit a pull request you have already tested it or warned me if it is not complete.

Ok. Understood.

I have tested  this submission and saw no error nor strange behaviours. I created a script that use the new API methods and it worked flawlessly.

I will add the other methods to the API so it supports the other properties supported by org.freeplane.features.nodestyle.NodeBorderModel.

While working and modifying this branch, should I close this pull request and reopen it when ready? Or create a new pull request?

How can I "pause" this pull request, so I am sure nobody will accidentally merge it to the main branch?

Sorry about the questions, but I have only worked alone in my repos in Github, so I don't know how is the protocol in these cases.

thanks,

dpolivaev commented:
Thank you for asking, the protocol depends on the project.
The pull request remains open until we both agree that it can be accepted or should be rejected.
Fixing related issues is a usual part of the pull request life cycle.

EdoFro committed:
simplifying expressions

EdoFro committed:
Adding width methods to Border api

dpolivaev commented:
NodeBorderModel has more properties:

`````
	private Boolean borderWidthMatchesEdgeWidth = null;
	private Quantity<LengthUnit> borderWidth = null;

	private Boolean borderDashMatchesEdgeDash = null;
	private DashVariant borderDash = null;

	private Boolean borderColorMatchesEdgeColor = null;
	private Color borderColor = null;
`````

Why have you not added all of them to the API? Why have you added the isSet method only for Color but not for the other properties?

EdoFro commented:
Hi Dimitry,
I had problems trying to implement `borderDash` property. I don't know how to create the `DashVariant` api.

I created `freeplane_api/src/main/java/org/freeplane/api/DashVariant.java` as `public enum` but I don't know how to relate it to `freeplane/src/main/java/org/freeplane/features/DashVariant.java`.

Can You give me some indication what should I do?

Thanks,

edo

dpolivaev commented:
I would refactor the code so that the enum does not need to implement RenderedContentSupplier and can be moved to Freeplane API. I would introduce a new method 

`````
public static <T extends Enum<T>> ComboProperty of(String name, Class<T> enumClass, Function<T, RenderedContent<T>) 
````` 

so that the mapping currently implemented inside the enum can be implemented where the function is called (ConnectorDashControlGroup) or somewhere else.

I can do the refactoring and send you a pull request.

EdoFro commented:
Yes, please.
No hurry. This weekend I will have no time for this. 
Maybe on monday can I try it again.

Thanks. Have a nice weekend.

edo

dpolivaev committed:
Move RenderedContentSupplier methods from enums so that the enums can be used in freeplane api

EdoFro committed:
Merge pull request #2 from freeplane/border-script-api

Move RenderedContentSupplier methods from enums

dpolivaev commented:
Hello Edo, can you please update me on the status of this pull request? I did some preparatory work which could allow you moving the enums to the API package, but I have neither actually moved them nor added the new methods to the API. Are you still going to do it?

EdoFro commented:
Hello Dimitry,
Yes, I´m interested in doing this. I worked last week in other project in my free time, but this week I want to continue with this.

Thank you for your support,

Edo

EdoFro commented:
Hi DImitry,
I moved DashVariant.java to the API using IntelliJ's "move Class" refactoring option. It seems it worked ok, because all classes in the main project that use this class, now have the `import org.freeplane.api.DashVariant` sentence in their code.

But I have a new question:
Now **DashVariant.java** has no access to UITools.

![image](https://github.com/freeplane/freeplane/assets/35700575/49b8e407-e2e3-49b1-a2fa-614e3bd924eb)

and I don't know if I have to add that dependency on the module. Should I?

The class uses one method and one constant from UITools:
- UITools.createStroke(...)
- UITools.FONT_SCALE_FACTOR

EdoFro commented:
I will try to create a DashVariant enum in the API and extend it in Main.

It's that the way of solving this?

dpolivaev commented:
I would rather move the static methods to a class defined in the api package and keep methods in UITools as delegates so that the dependency on UI tools is not needed.

EdoFro commented:
Ok. I understand.

I'm trying now to implement it using the Edge API as example. 
I want to see if it works. After that I will try what you suggest.

EdoFro commented:
I tried the whole afternoon to get this working and I couldn't. It is almost 22:00 and I can't understand what I'm doing wrong or what is the correct way to implement this.

I wanted to implement the borderDash setter and getter in the API to have a way to read the style of a node and make an UI button simulating that style. For example, to implement a tab in the TabbedPanel with buttons to apply the styles in the nodes.

I don't know if I will try again. It's way over my coding knowledge.

EdoFro commented:
The properties that we have already implemented are also useful for me. 

I created a script that "shifts" the colors in the styles map.

for example this map:
![image](https://github.com/freeplane/freeplane/assets/35700575/8e78b184-5d3f-4046-80cc-dfdacb722027)

can be "restyled" like this:
![image](https://github.com/freeplane/freeplane/assets/35700575/c5f759bd-0a1e-454c-814a-3a51490cb598)


I was thinking in transform it into an addon, and that's why I was trying to implement this changes in the API.

EdoFro committed:
Merge branch 'freeplane:1.11.x' into test_edo_Api_changes

dpolivaev committed:
Merge branch '1.11.x' into edo-api

dpolivaev committed:
Decouple DashVariant from icons and UITools

dpolivaev committed:
Apply font scale factor to dashed curves

dpolivaev committed:
Move DashVariant to freeplane api and improve naming

dpolivaev committed:
implement new API methods

dpolivaev commented:
Hello Edo, could you please test how it works with my recent changes? If you find it is ok I will merge it.

EdoFro committed:
Merge branch 'freeplane:1.11.x' into test_edo_Api_changes

EdoFro commented:
Hi Dimitry,
I will test it this weekend.

I had a first error message when opening a mindMap: "**Can not process attribute BORDER_DASH = 'CLOSE_DOTS'** "

<details> <summary> Log </summary>

```
STDERR: Oct 27, 2023 9:05:20 PM org.freeplane.core.util.LogUtils severe
SEVERE: Can not process attribute BORDER_DASH = 'CLOSE_DOTS'
java.lang.IllegalArgumentException: No enum constant org.freeplane.api.Dash.CLOSE_DOTS
	at java.base/java.lang.Enum.valueOf(Enum.java:273)
	at org.freeplane.api.Dash.valueOf(Dash.java:7)
	at org.freeplane.features.nodestyle.NodeStyleBuilder$26.setAttribute(NodeStyleBuilder.java:323)
	at org.freeplane.core.io.xml.TreeXmlReader.addAttribute(TreeXmlReader.java:110)
	at org.freeplane.core.io.xml.TreeXmlReader.addAttribute(TreeXmlReader.java:129)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElement(StdXMLParser.java:293)
	at org.freeplane.core.io.xml.XMLParser.processElement(XMLParser.java:47)
	at org.freeplane.n3.nanoxml.StdXMLParser.scanSomeTag(StdXMLParser.java:489)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElementContent(StdXMLParser.java:363)
	at org.freeplane.core.io.xml.XMLParser.processElementContent(XMLParser.java:153)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElement(StdXMLParser.java:315)
	at org.freeplane.core.io.xml.XMLParser.processElement(XMLParser.java:47)
	at org.freeplane.n3.nanoxml.StdXMLParser.scanSomeTag(StdXMLParser.java:489)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElementContent(StdXMLParser.java:363)
	at org.freeplane.core.io.xml.XMLParser.processElementContent(XMLParser.java:153)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElement(StdXMLParser.java:315)
	at org.freeplane.core.io.xml.XMLParser.processElement(XMLParser.java:47)
	at org.freeplane.n3.nanoxml.StdXMLParser.scanSomeTag(StdXMLParser.java:489)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElementContent(StdXMLParser.java:363)
	at org.freeplane.core.io.xml.XMLParser.processElementContent(XMLParser.java:153)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElement(StdXMLParser.java:315)
	at org.freeplane.core.io.xml.XMLParser.processElement(XMLParser.java:47)
	at org.freeplane.n3.nanoxml.StdXMLParser.scanSomeTag(StdXMLParser.java:489)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElementContent(StdXMLParser.java:363)
	at org.freeplane.core.io.xml.XMLParser.processElementContent(XMLParser.java:153)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElement(StdXMLParser.java:315)
	at org.freeplane.core.io.xml.XMLParser.processElement(XMLParser.java:47)
	at org.freeplane.n3.nanoxml.StdXMLParser.scanSomeTag(StdXMLParser.java:489)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElementContent(StdXMLParser.java:363)
	at org.freeplane.core.io.xml.XMLParser.processElementContent(XMLParser.java:153)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElement(StdXMLParser.java:315)
	at org.freeplane.core.io.xml.XMLParser.processElement(XMLParser.java:47)
	at org.freeplane.n3.nanoxml.StdXMLParser.scanSomeTag(StdXMLParser.java:489)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElementContent(StdXMLParser.java:363)
	at org.freeplane.core.io.xml.XMLParser.processElementContent(XMLParser.java:153)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElement(StdXMLParser.java:315)
	at org.freeplane.core.io.xml.XMLParser.processElement(XMLParser.java:47)
	at org.freeplane.n3.nanoxml.StdXMLParser.scanSomeTag(StdXMLParser.java:489)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElementContent(StdXMLParser.java:363)
	at org.freeplane.core.io.xml.XMLParser.processElementContent(XMLParser.java:153)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElement(StdXMLParser.java:315)
	at org.freeplane.core.io.xml.XMLParser.processElement(XMLParser.java:47)
	at org.freeplane.n3.nanoxml.StdXMLParser.scanSomeTag(StdXMLParser.java:489)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElementContent(StdXMLParser.java:363)
	at org.freeplane.core.io.xml.XMLParser.processElementContent(XMLParser.java:153)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElement(StdXMLParser.java:315)
	at org.freeplane.core.io.xml.XMLParser.processElement(XMLParser.java:47)
	at org.freeplane.n3.nanoxml.StdXMLParser.scanSomeTag(StdXMLParser.java:489)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElementContent(StdXMLParser.java:363)
	at org.freeplane.core.io.xml.XMLParser.processElementContent(XMLParser.java:153)
	at org.freeplane.n3.nanoxml.StdXMLParser.processElement(StdXMLParser.java:315)
	at org.freeplane.core.io.xml.XMLParser.processElement(XMLParser.java:47)
	at org.freeplane.n3.nanoxml.StdXMLParser.scanSomeTag(StdXMLParser.java:489)
	at org.freeplane.n3.nanoxml.StdXMLParser.scanData(StdXMLParser.java:447)
	at org.freeplane.n3.nanoxml.StdXMLParser.parse(StdXMLParser.java:120)
	at org.freeplane.core.io.xml.TreeXmlReader.load(TreeXmlReader.java:257)
	at org.freeplane.core.io.xml.TreeXmlReader.load(TreeXmlReader.java:334)
	at org.freeplane.features.map.MapReader$NodeTreeCreator.create(MapReader.java:62)
	at org.freeplane.features.map.MapReader$NodeTreeCreator.createNodeTreeFromXml(MapReader.java:76)
	at org.freeplane.features.map.MapReader.createNodeTreeFromXml(MapReader.java:170)
	at org.freeplane.features.map.MapReader.createNodeTreeFromXml(MapReader.java:163)
	at org.freeplane.features.url.mindmapmode.MFileManager.loadTreeImpl(MFileManager.java:563)
	at org.freeplane.features.url.mindmapmode.MFileManager.loadTree(MFileManager.java:533)
	at org.freeplane.features.url.mindmapmode.MapLoader.loadMap(MapLoader.java:294)
	at org.freeplane.features.url.mindmapmode.MapLoader.access$200(MapLoader.java:35)
	at org.freeplane.features.url.mindmapmode.MapLoader$1.run(MapLoader.java:230)
	at org.freeplane.features.url.mindmapmode.MapLoader$1.run(MapLoader.java:221)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at org.freeplane.features.url.mindmapmode.MapLoader.loadMap(MapLoader.java:221)
	at org.freeplane.features.url.mindmapmode.MapLoader.getMap(MapLoader.java:134)
	at org.freeplane.features.map.mindmapmode.MMapController.openMap(MMapController.java:1040)
	at org.freeplane.features.url.UrlManager.loadOtherHyperlink(UrlManager.java:349)
	at org.freeplane.features.url.UrlManager.loadHyperlink(UrlManager.java:281)
	at org.freeplane.features.url.mindmapmode.MFileManager.loadHyperlink(MFileManager.java:603)
	at org.freeplane.features.link.LinkController.loadHyperlink(LinkController.java:583)
	at org.freeplane.features.link.LinkController.loadURI(LinkController.java:1093)
	at org.freeplane.features.link.mindmapmode.MLinkController.loadURI(MLinkController.java:1001)
	at org.freeplane.features.link.LinkController.loadURL(LinkController.java:623)
	at org.freeplane.features.link.LinkController.loadURL(LinkController.java:592)
	at org.freeplane.features.link.mindmapmode.MLinkController.loadURL(MLinkController.java:983)
	at org.freeplane.features.link.FollowLinkAction.actionPerformed(FollowLinkAction.java:44)
	at java.desktop/javax.swing.SwingUtilities.notifyAction(SwingUtilities.java:1810)
	at org.freeplane.core.ui.ActionAcceleratorManager.processKeyBinding(ActionAcceleratorManager.java:454)
	at org.freeplane.main.mindmapmode.MModeControllerFactory$2.processKeyBinding(MModeControllerFactory.java:291)
	at org.freeplane.view.swing.ui.UserInputListenerFactory$2.processKeyBinding(UserInputListenerFactory.java:176)
	at org.freeplane.core.ui.components.FreeplaneMenuBar.processKeyBinding(FreeplaneMenuBar.java:102)
	at org.freeplane.main.application.MenuKeyProcessor.processKeyBinding(MenuKeyProcessor.java:26)
	at org.freeplane.main.application.ConnectedToMenuView.processKeyBinding(ConnectedToMenuView.java:41)
	at java.desktop/javax.swing.JComponent.processKeyBindings(JComponent.java:3009)
	at java.desktop/javax.swing.JComponent.processKeyEvent(JComponent.java:2909)
	at java.desktop/java.awt.Component.processEvent(Component.java:6403)
	at java.desktop/java.awt.Container.processEvent(Container.java:2266)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5001)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4833)
	at java.desktop/java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1952)
	at java.desktop/java.awt.Defaul
STDERR: tKeyboardFocusManager.dispatchKeyEvent(DefaultKeyboardFocusManager.java:883)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.preDispatchKeyEvent(DefaultKeyboardFocusManager.java:1150)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:1020)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:848)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:4882)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2780)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4833)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:773)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:722)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:716)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:746)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:744)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:743)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

```

</details>

Regards,

Edo

dpolivaev committed:
Use the old name to ensure backward compatibility with old maps

dpolivaev commented:
I am sorry, I fixed the CLOSE_DOTS issue. My bad.

EdoFro commented:
I decided to add the other methods like: borderColorMatchesEdgeColor

till now everything works fine

dpolivaev commented:
Great, I am looking forward to it. Please let me know when you think I could merge everything.

EdoFro committed:
corrects border.isWithSet

EdoFro committed:
implements "border Matches Edge ..." methods

EdoFro committed:
implements "isSet" for "border Matches Edge..." in Border API

implements:
- border.isUsesEdgeColorSet()
- border.isUsesEdgeWidthSet()
- border.isUsesEdgeDashSet()

EdoFro committed:
implements additional border.setWidth() methods

EdoFro committed:
Adds some explanation for API's java doc

EdoFro commented:
@dpolivaev,
I don't know if I did something wrong. Please take care before merging.

Explanation:
I added 5 new commits today. They are ok. I have tested them. (the last one is "Adds some explanation for API's java doc"

but then it appeared a message saying that there were conflicts in the LinkController.java file with some changes that you made recently, and I tried to solved them in my computer, but I clicked something and I don't know what it did. I tried to revert it but now there is no conflict anymore, but I don't know what or if something changed.

EdoFro commented:
I have investigating a little more and I DID SOMETHING WRONG.

The last two commits are wrong.

I won't touch anything more, because I don't want to make it worse. I hope you can solve it.

Sorry for the inconvenience.

Edo

dpolivaev committed:
Merge remote-tracking branch 'origin/1.11.x' into test_edo_Api_changes

dpolivaev commented:
I reverted the branch to the last commit before your merge attempt and pushed my merged version.
Now you need to execute something like

````
git fetch
git reset --hard a2f5326dbd83c6482624a734f6777f388bc2752c
````

dpolivaev commented:
(if you named your repo  remote as the origin, you can alternatively run

````
git fetch
git reset --hard origin/test_edo_Api_changes
````
)

EdoFro commented:
done. thanks!

EdoFro commented:
tested. it works perfect.
I think it can be merged.

👍🏼

dpolivaev commented:
The new API is available in preview 1.11.8_06
